integration/tests/common/block_state.go

```diff
@@ -23,61 +23,75 @@ type BlockState struct {
 	highestSealed     *messages.BlockProposal
 }
 
-// TODO refactor to remove deep indentation
+func NewBlockState() *BlockState {
+	return &BlockState{
+		Mutex:             sync.Mutex{},
+		blocksByID:        make(map[flow.Identifier]*messages.BlockProposal),
+		finalizedByHeight: make(map[uint64]*messages.BlockProposal),
+	}
+}
+
 func (bs *BlockState) Add(b *messages.BlockProposal) {
 	bs.Lock()
 	defer bs.Unlock()
 
-	if bs.blocksByID == nil {
-		bs.blocksByID = make(map[flow.Identifier]*messages.BlockProposal) // TODO: initialize this map in constructor
-	}
 	bs.blocksByID[b.Header.ID()] = b
-	bs.highestProposed = b.Header.Height
+	if b.Header.Height > bs.highestProposed {
+		bs.highestProposed = b.Header.Height
+	}
 
-	// add confirmations
-	confirmsHeight := b.Header.Height - 3
-	if b.Header.Height >= 3 && confirmsHeight > bs.highestFinalized {
-		if bs.finalizedByHeight == nil {
-			bs.finalizedByHeight = make(map[uint64]*messages.BlockProposal) // TODO: initialize this map in constructor
-		}
-		// put all ancestors into `finalizedByHeight`
-		ancestor, ok := b, true
-		for ancestor.Header.Height > bs.highestFinalized {
-			h := ancestor.Header.Height
+	if b.Header.Height < 3 {
+		return
+	}
+
+	confirmsHeight := b.Header.Height - uint64(3)
+	if confirmsHeight < bs.highestFinalized {
+		return
+	}
 
-			// if ancestor is confirmed put it into the finalized map
-			if h <= confirmsHeight {
+	bs.processAncestors(b, confirmsHeight)
+}
 
-				finalized := ancestor
-				bs.finalizedByHeight[h] = finalized
+func (bs *BlockState) processAncestors(b *messages.BlockProposal, confirmsHeight uint64) {
+	// puts this block proposal and all ancestors into `finalizedByHeight`
+	ancestor, ok := b, true
+	for ancestor.Header.Height > bs.highestFinalized {
+		h := ancestor.Header.Height
 
-				// if ancestor confirms a heigher height than highestFinalized, increase highestFinalized
-				if h > bs.highestFinalized {
-					bs.highestFinalized = h
-				}
+		// if ancestor is confirmed put it into the finalized map
+		if h <= confirmsHeight {
 
-				// update last sealed height
-				for _, seal := range finalized.Payload.Seals {
-					sealed, ok := bs.blocksByID[seal.BlockID]
-					if !ok {
-						continue
-					}
-
-					if bs.highestSealed == nil ||
-						sealed.Header.Height > bs.highestSealed.Header.Height {
-						bs.highestSealed = sealed
-					}
-				}
+			finalized := ancestor
+			bs.finalizedByHeight[h] = finalized
 
+			// highestFinalized is only updated when not only the block is added to finalizedByHeight, but also
+			// it is "exactly" the next height compared to the current value.
+			// doing so prevents the illusion of highestFinalized indicating the highest finalized height "with an available block".
+			if h == bs.highestFinalized+1 {
```

Comment:
I'm worried that this logic can cause the `highestFinalized` to lag permanently behind because we are looping from high to low heights. Suppose there is a gap in the blocks (block 10 is missing), then we receive blocks 11-20. The `highestFinalized` block height must be less than 10 while the latest unfinalized height is 20. When we receive block 10, and each subsequent block, we will increase the `highestFinalized` value by at most 1 because of this condition and the iteration order. So the gap between latest unfinalized and latest finalized will remain >=10.

Reply:
Addressed this comment by separating the logic of processing ancestors (high-to-low) and updating the highest finalized height (low-to-heigh) in https://github.com/onflow/flow-go/pull/936/commits/7af850e3bae1763959435656efaecd19eed6abf4.